sherpa/plot/bokeh_backend.py

```diff
@@ -0,0 +1,829 @@
+        """
+        x0 = np.unique(x0)
+        x1 = np.unique(x1)
+        y = np.asarray(y)
+
+        if x0.size * x1.size != y.size:
+            raise NotImplementedErr('contourgrids')
+
+        y = y.reshape(x1.size, x0.size)
+
+        axes = self.setup_axes(overcontour, clearwindow)
+
+        # Set up the axes
+        if not overcontour:
+            self.setup_plot(axes, title, xlabel, ylabel, xlog=xlog, ylog=ylog)
+
+        if levels is None:
+            levels = 7
+        if isinstance(levels, int):
+            # Matplotlib has defaults for levels and sherpa can pass in "None" to mean
+            # "use the default". Bokeh always requires to specify the levels, so
+            # we have to define our own defaults here.
+            # We try to pick something sensible, but not too complicated.
+            # There are cases, where this simple appraoch does not
+            # provide nice nubmers for the levels, but matplotlib's
+            # MaxNLocator is > 200 lines, which I don't want to replicate.
+            levels = np.linspace(np.min(y), np.max(y), levels)[1: -1]
+
+
+        axes.contour(x0, x1, y, levels, line_alpha=alpha,
+                     line_color=colors, line_width=linewidths)
+
+    def _index_axis(self, row, col):
+        '''Find index number of subplot in a grid of plots
+
+        self.current_fig is a `bokeh.models.plots.GridPlot`
+        object. It has a children attribute, which is a list of tuples, where
+        each tuple is a (figure, row, col) tuple. This method returns the
+        index of the tuple in the list that matches the given row and col or,
+        if no such tuple exists, None.
+        '''
+        for i, fig in enumerate(self.current_fig.children):
+            if fig[1] == row and fig[2] == col:
+                return i
+        return None
+
+    def set_subplot(self, row, col, nrows, ncols, clearaxes=True,
+                    left=None,
+                    right=None,
+                    bottom=None,
+                    top=None,
+                    wspace=0.3,
+                    hspace=0.4):
+        """Select a plot space in a grid of plots or create new grid
+
+        This method adds a new subplot in a grid of plots.
+
+        Parameters
+        ----------
+        row, col : int
+            index (starting at 0) of a subplot in a grid of plots
+        nrows, ncols : int
+            Number of rows and column in the plot grid
+        clearaxes : bool
+            If True, clear entire plotting area before adding the new
+            subplot.
+        """
+        index = self._index_axis(row, col)
+        if index is not None and not clearaxes:
+            # We found the right subplot, now make it the current one
+            self.current_axis = self.current_fig[index]
+            return
+
+        # We found the subplot, but we want to replace it with a new one
+        if index is not None and clearaxes:
+            self.current_fig.children.pop(index)
```

Comment:
So, pylint is complainnig to me that "Instance of 'List' has no 'pop' member". This strikes me as odd, because a list does have pop, but perhaps whatever `children` is does not.

There is some strangeness here, since in line 658 above you have `current_fig` as being a sequence (i.e has an `[index]` element) but here `current_fig` also has a `children` argument. Both are entirely possible in the wild world of Python, but it's not clear whether we test this (I don't entirely trust the "line not covered by test" lines from codecov, but I can't realistically test this at the moment to check [*]).

[*] I am old school and stick an `assert False` at the location and re-test

Reply:
I did stick in `assert False` and it does make some tests fail, so this code is used.

Reply:
I think I put the `assert False` in the wrong location and there is a "children" missing above, but in a code path only triggered to multi-panel plots where individual panels are re-created. We don't have tests to cover use (or use cases) and I don't trust the pylab backend either for that...